internal/util/flowgraph/node.go

```diff
@@ -116,50 +98,158 @@ func (nodeCtx *nodeCtx) work() {
 
 	for {
 		select {
-		case <-nodeCtx.closeCh:
-			log.Debug("flow graph node closed", zap.String("nodeName", nodeCtx.node.Name()))
+		case <-nodeCtxManager.inputNodeCloseCh:
 			return
+		// handles node work spinning
+		// 1. collectMessage from upstream or just produce Msg from InputNode
+		// 2. invoke node.Operate
+		// 3. deliver the Operate result to downstream nodes
 		default:
 			// inputs from inputsMessages for Operate
 			var input, output []Msg
-			if !nodeCtx.node.IsInputNode() {
-				input = <-nodeCtx.inputChannel
-			}
+			// inputNode.input not from nodeCtx.inputChannel
 			// the input message decides whether the operate method is executed
-			n := nodeCtx.node
-			nodeCtx.blockMutex.RLock()
+			n := inputNode.node
+			inputNode.blockMutex.RLock()
 			if !n.IsValidInMsg(input) {
-				nodeCtx.blockMutex.RUnlock()
+				inputNode.blockMutex.RUnlock()
 				continue
 			}
 			output = n.Operate(input)
-			nodeCtx.blockMutex.RUnlock()
+			inputNode.blockMutex.RUnlock()
 			// the output decide whether the node should be closed.
 			if isCloseMsg(output) {
-				close(nodeCtx.closeCh)
-				nodeCtx.closeWg.Done()
-				nodeCtx.node.Close()
-				if nodeCtx.inputChannel != nil {
-					close(nodeCtx.inputChannel)
+				close(nodeCtxManager.inputNodeCloseCh)
+				// inputNode.Close()
+				if inputNode.inputChannel != nil {
+					close(inputNode.inputChannel)
 				}
 			}
-
+			// deliver to all following flow graph node.
+			inputNode.downstream.inputChannel <- output
 			if enableTtChecker {
 				checker.Check(name)
 			}
+		}
+	}
+}
 
-			// deliver to all following flow graph node.
-			if nodeCtx.downstream != nil {
-				nodeCtx.downstream.inputChannel <- output
+func (nodeCtxManager *nodeCtxManager) ddNodeWork() {
```

Comment:
ditto, ddNode is a  node named and created by DataNode, not very suitable for flowgraph.node